[FLINK-6091] [table] Implement and turn on retraction for aggregates

hequn8128 opened the pull request:
Implement functions for generating and consuming retract messages for different aggregates.

1. add delete/add property to Row
2. implement functions for generating retract messages for unbounded groupBy
3. implement functions for handling retract messages for different aggregates.
4. handle retraction messages in `CommonCorrelate` and `CommonCalc` (retain Delete property).

Currently, only unbounded groupby generates retraction and it is working under unbounded and processing time mode. Hence, so far retraction is only supported for unbounded and processing time aggregations. We can add more retraction support later.

supported now: unbounded groupby, unbounded and processing time over window
unsupported now: group window, event time or bounded over window.

lincoln-lil commented on flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/RowSerializer.java:
```diff
@@ -40,6 +42,7 @@
 	private static final long serialVersionUID = 1L;
 	private final boolean[] nullMask;
 	private final TypeSerializer<Object>[] fieldSerializers;
+	private final EnumSerializer commandSerializer = new EnumSerializer(Command.class);
```
advice: non-static final members should be initialized in constructor, otherwise make it 'static final' and use upper case variable name.

lincoln-lil commented on flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/RowSerializer.java:
```diff
@@ -171,6 +181,7 @@ public Row deserialize(DataInputView source) throws IOException {
 				result.setField(i, fieldSerializers[i].deserialize(source));
 			}
 		}
+		result.command = (Command) commandSerializer.deserialize(source);
```
should we use the method with reuse variable ?
T deserialize(T reuse, DataInputView source)
though it doesn't perform reusing actually

lincoln-lil commented on flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/RowSerializer.java:
```diff
@@ -198,6 +209,7 @@ public Row deserialize(Row reuse, DataInputView source) throws IOException {
 				}
 			}
 		}
+		reuse.command = (Command) commandSerializer.deserialize(source);
```
same with L184

lincoln-lil commented on flink-core/src/main/java/org/apache/flink/types/Command.java:
```diff
@@ -0,0 +1,29 @@
+
+/**
+ * A Command is used in a {@link Row} to distinguish delete or add.
+ * Delete indicate a retracion row and Add means a nomal row.
```
retracion -> retraction

lincoln-lil commented on flink-core/src/main/java/org/apache/flink/types/Row.java:
```diff
@@ -46,6 +46,9 @@
 	/** The array to store actual values. */
 	private final Object[] fields;
 
+	/** Indicate to add or delete this row */
```
I prefer "Indicates that the line is add or delete type"

lincoln-lil commented on flink-core/src/main/java/org/apache/flink/types/Row.java:
```diff
@@ -46,6 +46,9 @@
 	private final Object[] fields;
 
+	/** Indicate to add or delete this row */
+	public Command command = Command.Add;
```
I think change to private and offer getter/setter will be more better.

lincoln-lil commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/CommonCalc.scala:
```diff
@@ -52,10 +53,18 @@ trait CommonCalc {
       rowType.getFieldNames,
       expandedExpressions)
 
+    val retractionProcess =
```
rename retractionProcess to retractionProcessCode
or moving this part to CodeGenerator considering the consistency of code style.
And I prefer the later one.

lincoln-lil commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/CommonCorrelate.scala:
```diff
@@ -135,11 +136,20 @@ trait CommonCorrelate {
       val outerResultExpr = functionGenerator.generateResultExpression(
         input1AccessExprs ++ input2NullExprs, returnType, rowType.getFieldNames.asScala)
+
+      val retractionProcess =
```
see CommonCalc.scala: L56

lincoln-lil commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/CommonCorrelate.scala:
```diff
@@ -182,9 +192,17 @@ trait CommonCorrelate {
       returnType,
       rowType.getFieldNames.asScala)
 
+    val retractionProcess =
```
see CommonCalc.scala: L56

lincoln-lil commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamGroupWindowAggregate.scala:
```diff
@@ -107,6 +107,14 @@ class DataStreamGroupWindowAggregate(
+    if (consumeRetraction) {
+      throw new TableException(
+        "Retraction on group window is not supported yet. Note: Currently, group window should " +
+          "not follow an unbounded groupby.")
```
1. 'is not supported yet' -> 'unsupported yet' maybe more concisely.
2. groupby -> groupBy

hequn8128 replied:
yes, non-windowed is more clear

lincoln-lil commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamOverAggregate.scala:
```diff
@@ -129,11 +131,17 @@ class DataStreamOverAggregate(
           // bounded OVER window
+          if (consumeRetraction) {
+            throw new TableException(
+              "Retraction for bounded over window is not supported yet. Note: Currently, bounded " +
```
1. Retraction for -> on
2. groupby -> groupBy
3. keep the term consistent either choosing 'OVER' or 'over' in the same context.

lincoln-lil commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/GroupAggProcessFunction.scala:
```diff
@@ -68,12 +70,35 @@ class GroupAggProcessFunction(
+        }
+        i = 0
+        while (i < aggregates.length) {
+          val index = groupings.length + i
```
reduce a local variable:
previous.setField(groupings.length + i, aggregates(i).getValue(accumulator))

also L118 and L127

lincoln-lil commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/GroupAggProcessFunction.scala:
```diff
@@ -84,16 +109,38 @@ class GroupAggProcessFunction(
+    if (null != previous) {
+      if (previous.equals(output)) {
+        // ignore same output
+        return
```
Is it possible which needs output the 'same' row ?

fhueske replied:
We still need to update the state even if we do not emit new values. If we have a max aggregation with an accumulator that holds a map of `10->1, 5->2` and we add `8`. The new accumulator will be `10->1, 8->1, 5->2` but the aggregation result will still be 10. If we later retract `10`, the new max would be `5` but should be `8`.

hequn8128 replied:
hi, we need not to output a same row, think about batch, only one result will be output, is it right?

hequn8128 replied:
yes, you are right, we should update the state even if we do not emit new values. thx

lincoln-lil commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/scala/stream/RetractionITCase.scala:
```diff
@@ -0,0 +1,310 @@
+    results.addSink(new StreamITCase.StringSink)
+    env.execute()
+
+    val expected = Seq("1", "2", "1", "3", "4")
```
Could you explain this example?
table.groupBy('word).select('word as 'word, 'num.sum as 'count)
I think this query does not produce retraction

hequn8128 replied:
hi, there is a `.select('count.sum)` after `groupBy('word).select('word as 'word, 'num.sum as 'count)`

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/GroupAggProcessFunction.scala:
```diff
@@ -68,12 +70,35 @@ class GroupAggProcessFunction(
+      // get previous row
+      if (generateRetraction) {
+        if (null == previous) {
+          previous = new Row(groupings.length + aggregates.length)
```
can we initialize `previous` in `open()`?

hequn8128 replied:
done

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/GroupAggProcessFunction.scala:
```diff
@@ -41,14 +41,16 @@ class GroupAggProcessFunction(
   Preconditions.checkNotNull(aggFields)
   Preconditions.checkArgument(aggregates.length == aggFields.length)
 
   private var output: Row = _
```
rename to `newRow`.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/GroupAggProcessFunction.scala:
```diff
@@ -41,14 +41,16 @@ class GroupAggProcessFunction(
   Preconditions.checkArgument(aggregates.length == aggFields.length)
 
   private var output: Row = _
+  private var previous: Row = _
```
rename to `previousRow`.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/GroupAggProcessFunction.scala:
```diff
@@ -84,16 +109,38 @@ class GroupAggProcessFunction(
-    state.update(accumulators)
 
+    // if previous is not null, do retraction process
+    if (null != previous) {
```
check against `generateRetraction`. The check can be optimized because `generateRetraction`is a `val` and hence `final`.

fhueske commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/scala/stream/RetractionITCase.scala:
```diff
@@ -0,0 +1,310 @@
+  */
+class RetractionITCase extends StreamingWithStateTestBase {
+  // input data
+  val data = List(
```
Please use more test data.

fhueske commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/scala/stream/RetractionITCase.scala:
```diff
@@ -0,0 +1,310 @@
+  def testUniqueProcess(): Unit = {
+    // data input
+    val data = List(
+      (1234, 2L),
```
Please use more test data

fhueske commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/scala/stream/RetractionITCase.scala:
```diff
@@ -0,0 +1,310 @@
+
+  // groupby + window agg
+  @Test(expected = classOf[TableException])
+  def testGroupByAndProcessingTimeSlidingGroupWindow(): Unit = {
```
Please check the tests with expected failures to a test that extends `TableTestBase` such as `GroupAggregationsTest`.

hequn8128 replied:
hi, these failures are throwed during `translateToPlan`,  so `TableTestBase` can not help, right ?

hequn8128 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamOverAggregate.scala:
```diff
@@ -129,11 +131,17 @@ class DataStreamOverAggregate(
             isRowTimeType = false,
-            isRowsClause = overWindow.isRows)
+            isRowsClause = overWindow.isRows,
+            consumeRetraction)
```
hi, you are right, maybe it's better for us not to support OVER at current.

hequn8128 commented:
Hi @fhueske , the latest commit is a bugfix. It mainly includes the following contents:

1. request type add `CRow` support in `TableEnvironment`，so we can sink with `CRow`
2. `getFieldInfo` add `CRow` support in `TableEnvironment`，so we can init a `Table` from a `DataStream` of `CRow`
3. fix type erasure problem in `TableEnvironment` and `DataStreamRetractionRules`

Thanks, Hequn

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/TableEnvironment.scala:
```diff
@@ -627,24 +640,27 @@ abstract class TableEnvironment(val config: TableConfig) {
     */
-  protected def sinkConversion[T](
-      physicalRowTypeInfo: TypeInformation[Row],
+  protected def sinkConversion[T, P](
```
I would refactor this as follows:
- make `sinkConversion()` an abstract method that is implemented by `BatchTableEnvironemnt` and `StreamTableEnvironment`
- extract the code-generation part that converts a `Row` into the requested data type and keep this in a method `TableEnvironment.generateRowConverterFunction()`
- the implementations of `sinkConversion()` can use the `generateRowConverterFunction`

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/CodeGenUtils.scala:
```diff
@@ -233,6 +234,9 @@ object CodeGenUtils {
       case ri: RowTypeInfo =>
         ProductAccessor(index)
 
+      case cri: CRowTypeInfo =>
```
Revert these changes

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/CodeGenerator.scala:
```diff
@@ -255,17 +256,23 @@ class CodeGenerator(
     * @return A GeneratedAggregationsFunction
     */
-  def generateAggregations(
+  def generateAggregations[T](
```
revert all changes in `CodeGenerator`. The wrapping functions (such as `MapRunner`) can extract the `Row` from a `CRow` and call the generated function with that object.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/CommonCalc.scala:
```diff
@@ -23,26 +23,28 @@ import org.apache.calcite.rel.`type`.RelDataType
 trait CommonCalc {
 
-  private[flink] def functionBody(
+  private[flink] def functionBody[T](
```
I would follow a similar strategy as with the `TableEnvironment` sink conversion for the `CommonX` classes.
The `CommonX` class has a method to generate the code for a function that operates on `Row`. Each class that extends `CommonX` has a dedicated method to instantiate the wrapper class (e..g, `MapRunner`, `FlatMapRunner`. ). The runner classes use the generated function to operate on Rows. For streaming operators, the runner classes unwrap the `Row` from `CRow` before calling the generated code.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/dataset/BatchScan.scala:
```diff
@@ -38,7 +39,9 @@ trait BatchScan extends CommonScan with DataSetRel {
 
     val inputType = input.getType
 
-    val internalType = FlinkTypeFactory.toInternalRowTypeInfo(getRowType)
```
These changes can be reverted if we leave the `FlinkTypeFactory.toInternalRowTypeInfo()` method as it is.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamCalc.scala:
```diff
@@ -45,7 +45,7 @@ class DataStreamCalc(
   extends Calc(cluster, traitSet, input, calcProgram)
   with CommonCalc
-  with DataStreamRel {
+  with DataStreamRel[CRow] {
```
`DataStreamRel` does not need a type parameter, IMO. All `DataStreamRel` nodes should produce `CRow`.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamCalc.scala:
```diff
@@ -92,11 +92,13 @@ class DataStreamCalc(
 
     val genFunction = generator.generateFunction(
```
If we generate a `FlatMapFunction[Row, Row]` we can use a `RetractFlatMapRunner` class to unwrap the `Row` from the `CRow` before calling the generated code.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamRel.scala:
```diff
@@ -22,17 +22,16 @@ import org.apache.calcite.rel.RelNode
-import org.apache.flink.types.Row
 
-trait DataStreamRel extends RelNode with FlinkRel {
+trait DataStreamRel[T] extends RelNode with FlinkRel {
```
I don't think we need to parameterize `DataStreamRel`. All `DataStreamRel` nodes should return `CRow` in my opinion.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateAggFunction.scala:
```diff
@@ -46,12 +47,12 @@ class AggregateAggFunction(
 
-  override def add(value: Row, accumulatorRow: Row): Unit = {
+  override def add(value: CRow, accumulatorRow: Row): Unit = {
 
```
add here `val row = value.row` and use `row` to aggregate the values.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala:
```diff
@@ -61,23 +62,28 @@ object AggregateUtil {
     generator: CodeGenerator,
     namedAggregates: Seq[CalcitePair[AggregateCall, String]],
     inputType: RelDataType,
+    inputTypeInfo: TypeInformation[T],
```
We don't need these types if we keep the code generation as it is.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala:
```diff
@@ -102,13 +110,17 @@ object AggregateUtil {
           genFunction,
           aggregationStateType,
-          FlinkTypeFactory.toInternalRowTypeInfo(inputType))
+          FlinkTypeFactory
```
change to `CRowTypeInfo(FlinkTypeFactory.toInternalRowType(inputType)))`

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala:
```diff
@@ -102,13 +110,17 @@ object AggregateUtil {
           genFunction,
           aggregationStateType,
-          FlinkTypeFactory.toInternalRowTypeInfo(inputType))
+          FlinkTypeFactory
```
change to `CRowTypeInfo(FlinkTypeFactory.toInternalRowType(inputType)))`

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala:
```diff
@@ -118,7 +130,7 @@ object AggregateUtil {
         new ProcTimeUnboundedNonPartitionedOver(
           genFunction,
-          aggregationStateType)
+          new CRowTypeInfo(aggregationStateType))
```
we can keep `aggregationStateType` if we don't touch the code generation

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala:
```diff
@@ -159,18 +174,22 @@ object AggregateUtil {
     generator: CodeGenerator,
     namedAggregates: Seq[CalcitePair[AggregateCall, String]],
     inputType: RelDataType,
+    inputTypeInfo: TypeInformation[T],
```
remove new type info parameters.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala:
```diff
@@ -179,7 +198,9 @@ object AggregateUtil {
 
     val aggregationStateType: RowTypeInfo = createAccumulatorRowType(aggregates)
-    val inputRowType = FlinkTypeFactory.toInternalRowTypeInfo(inputType).asInstanceOf[RowTypeInfo]
+    val inputRowType = FlinkTypeFactory
```
change to `CRowTypeInfo(FlinkTypeFactory.toInternalRowType(inputType)))`

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSessionWindowAggReduceGroupFunction.scala:
```diff
@@ -66,7 +66,7 @@ class DataSetSessionWindowAggReduceGroupFunction(
   private var aggregateBuffer: Row = _
   private var output: Row = _
-  private var collector: TimeWindowPropertyCollector = _
+  private var collector: TimeWindowPropertyCollector[Row] = _
```
We can make `TimeWindowPropertyCollector` abstract and implement two versions, one for `Row` and one for `CRow`

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/GeneratedAggregations.scala:
```diff
@@ -31,35 +32,35 @@ abstract class GeneratedAggregations extends Function {
+    * @param output       output results collected in a command row
     */
-  def setAggregationResults(accumulators: Row, output: Row)
+  def setAggregationResults(accumulators: Row, output: CRow)
```
I would change all `CRow` types back to `Row`. Let's leave the code generation as it is and extract the rows in the wrapper classes.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/ProcTimeBoundedRangeOver.scala:
```diff
@@ -47,12 +48,12 @@ class ProcTimeBoundedRangeOver(
-  private var output: Row = _
+  private var output: CRow = _
   private var accumulatorState: ValueState[Row] = _
-  private var rowMapState: MapState[Long, JList[Row]] = _
```
I think we should keep the state as `Row`. At the moment we do not need the `command` field.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/ProcTimeBoundedRangeOver.scala:
```diff
@@ -69,10 +70,10 @@ class ProcTimeBoundedRangeOver(
     output = function.createOutputRow()
```
we can init `output` as `output = new CRow(function.createOutputRow(), true)`

fhueske replied:
When the code-gen'd functions need as `Row` to set the result. We can simply pass `output.row` then and later emit `output`.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/ProcTimeBoundedRangeOver.scala:
```diff
@@ -93,7 +94,7 @@ class ProcTimeBoundedRangeOver(
     // null value means that this si the first event received for this timestamp
     if (rowList == null) {
-      rowList = new ArrayList[Row]()
+      rowList = new ArrayList[CRow]()
```
Keep `Row`

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/ProcTimeBoundedRowsOver.scala:
```diff
@@ -48,15 +49,15 @@ class ProcTimeBoundedRowsOver(
   private var accumulatorState: ValueState[Row] = _
-  private var rowMapState: MapState[Long, JList[Row]] = _
-  private var output: Row = _
+  private var rowMapState: MapState[Long, JList[CRow]] = _
```
Same as for the other over window. Let's use `Row` for the state and extract the `Row` from the `CRow`.
`output` should be a `CRow`.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/ProcTimeBoundedRangeOver.scala:
```diff
@@ -82,9 +83,9 @@ class ProcTimeBoundedRangeOver(
+    ctx: ProcessFunction[CRow, CRow]#Context,
+    out: Collector[CRow]): Unit = {
 
     val currentTime = ctx.timerService.currentProcessingTime
```
we can set `val row = input.row` and add `row` instead of `input` to the state.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/TimeWindowPropertyCollector.scala:
```diff
@@ -19,34 +19,44 @@
   */
-class TimeWindowPropertyCollector(windowStartOffset: Option[Int], windowEndOffset: Option[Int])
-    extends Collector[Row] {
+class TimeWindowPropertyCollector[T](windowStartOffset: Option[Int], windowEndOffset: Option[Int])
```
we can make this an `abstract` class and add to subclasses `RowTimeWindowPropertyCollector` and `CRowTimeWindowPropertyCollector`.
That would avoid the `if` condition to identify the input type.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/types/CRow.scala:
```diff
@@ -0,0 +1,48 @@
+  }
+}
+
+class XRow(arity: Int, var change: Boolean) extends Row(arity) {
```
Can this be removed?

hequn8128 commented:
hi @fhueske , thanks a lot for your review and help. I have addressed all your comments and updated the PR. All changes have been checked before my latest update. 
Thanks,  Hequn

fhueske commented:
Thanks for the update @hequn8128. The PR looks really good, IMO.
Will do a couple of minor refactorings and merge it to `table-retraction` branch.

Best, Fabian

fhueske commented:
Hi @hequn8128, I merged the PR to `table-retraction`. Can you close it?

Thanks, Fabian
